DS/MainArrayList.java

```diff
@@ -0,0 +1,274 @@
+		if(location>=index) {
+			System.out.println("Please enter correct location");
+			return;
+		}
+		else {
+			int i=index-1;
+			while(i!=location) {
+				arrayList[i]=arrayList[i-1];
+				i--;
+			}
+			arrayList[location]=itemValue;
+		}
+	}
+	//this method is used to to find first location of user given item 
+	public void firstOccurrenceOfItem(int item) {
+		for(int i=0;i<index;i++) {
+			if(arrayList[i]==item) {
+				System.out.println("Item found at "+(i+1)+"location");
+				return;
+			}
+			else
+				System.out.println("Item not found");
+		}
+	}
+	//this method is used to to find first location of item at the location given by user
+	public void firstOccurrenceOfItemBasedOnLocation(int location) {
+		if(location>=index) {
+			System.out.println("Enter correct location");
+			return;
+		}
+		else {
+			int item=arrayList[location];
+			for(int i=0;i<index;i++) {
+				if(arrayList[i]==item) {
+					System.out.println("Item found at "+(i+1)+"location");
+					return;
+				}
+			}
+		}
+	}	
+	//this method is used to remove item from list at the location given by user 
+	public void removeItemAtGivenLocation(int location) {
+		if(location>=index) {
+			System.out.println("Please Enter correct Location:");
+			return;
+		}
+		else {
+			for(int i=location-1;i<index-1;i++) {
+				arrayList[i]=arrayList[i+1];
+			}
+			index--;
+		}
+			
+	}
+	//to remove first item given by user
+	public void removeItem(int item) {
+		boolean isItemFound=false;
+		for(int i=0;i<index;i++) {
+			if(arrayList[i]==item) {
+				for(int j=i;j<index-1;j++)
+					arrayList[j]=arrayList[j+1];
+					index--;
+					i--;
+					isItemFound=true;
+			}
+		}
+		if(isItemFound==false) {
+			System.out.println("Item not found");
+			return;
+		}
+	}
+	//to empty the list
+	public void removeAllElement() {
+		for(int i=0;i<index;i++) {
+			arrayList[i]=0;
+		}
+		index=0;
+	}
+	//to reverse the list
+	public void reverseList() {
+		int swapVariable;
+		for(int i=0;i<index/2;i++) {
+			swapVariable=arrayList[i];
+			arrayList[i]=arrayList[index-i-1];
+			arrayList[index-i-1]=swapVariable;
+		}
+	}
+	//to sort the list
+	public void toSortArrayList() {
+		for(int i=0;i<index;i++) {
+			for(int j=0;j<index-1-i;j++) {
+				if(arrayList[j]>arrayList[j+1]) {
+					int swapVariable;
+					swapVariable=arrayList[j];
+					arrayList[j]=arrayList[j+1];
+					arrayList[j+1]=swapVariable;
+				}
+			}
+		}
+	}
+	//this method is used to merge the two list
+	public void mergeList(ArrayList arrayListObject,ArrayList arrayListObject1) {
+		int newSize=arrayListObject.index+arrayListObject1.index;
+		int[] arrayListBuffer=new int[newSize];
+		int k=0;
+		for(int i=0;i<arrayListObject.index;i++)
+			arrayListBuffer[i]=arrayListObject.arrayList[i];
+		for(int i=arrayListObject.index;i<newSize;i++) {
+			arrayListBuffer[i]=arrayListObject1.arrayList[k++];
+			arrayListObject.index++;
+		}		
+		for(int i=0;i<arrayListObject.index;i++) {
```

Comment:
You can save compilation time by removing this loop from here and adding items in previous loops only.
It will save both, time complexity ( by eliminating loop ) and space complexity( as now we don't need arratListBuffer any more).